src/analysis/special_functions/trigonometric.lean

```diff
@@ -1478,88 +1478,332 @@ end
+    simp only [← cos_arcsin h₁.le h₂.le, one_div] at this ⊢,
+    exact ⟨sin_local_homeomorph.has_deriv_at_symm ⟨h₁, h₂⟩ this.ne' (has_deriv_at_sin _),
+      sin_local_homeomorph.times_cont_diff_at_symm_deriv this.ne' ⟨h₁, h₂⟩
+        (has_deriv_at_sin _) times_cont_diff_sin.times_cont_diff_at⟩ },
+  { have : 1 - x ^ 2 < 0, by nlinarith [h₂],
+    rw [sqrt_eq_zero'.2 this.le, div_zero],
+    have : arcsin =ᶠ[𝓝 x] λ _, π / 2 := (lt_mem_nhds h₂).mono (λ y hy, arcsin_of_one_le hy.le),
+    exact ⟨(has_deriv_at_const _ _).congr_of_eventually_eq this,
+      times_cont_diff_at_const.congr_of_eventually_eq this⟩ }
+end
+
+lemma has_deriv_at_arcsin {x : ℝ} (h₁ : x ≠ -1) (h₂ : x ≠ 1) :
+  has_deriv_at arcsin (1 / sqrt (1 - x ^ 2)) x :=
+(deriv_arcsin_aux h₁ h₂ 0).1
+
+lemma times_cont_diff_at_arcsin {x : ℝ} (h₁ : x ≠ -1) (h₂ : x ≠ 1) {n : with_top ℕ} :
+  times_cont_diff_at ℝ n arcsin x :=
+(deriv_arcsin_aux h₁ h₂ n).2
+
+lemma has_deriv_within_at_arcsin_Ici {x : ℝ} (h : x ≠ -1) :
+  has_deriv_within_at arcsin (1 / sqrt (1 - x ^ 2)) (Ici x) x :=
+begin
+  rcases em (x = 1) with (rfl|h'),
+  { convert (has_deriv_within_at_const _ _ (π / 2)).congr _ _;
+      simp [arcsin_of_one_le] { contextual := tt } },
+  { exact (has_deriv_at_arcsin h h').has_deriv_within_at }
+end
+
+lemma has_deriv_within_at_arcsin_Iic {x : ℝ} (h : x ≠ 1) :
+  has_deriv_within_at arcsin (1 / sqrt (1 - x ^ 2)) (Iic x) x :=
+begin
+  rcases em (x = -1) with (rfl|h'),
+  { convert (has_deriv_within_at_const _ _ (-(π / 2))).congr _ _;
+      simp [arcsin_of_le_neg_one] { contextual := tt } },
+  { exact (has_deriv_at_arcsin h' h).has_deriv_within_at }
+end
+
+lemma differentiable_within_at_arcsin_Ici {x : ℝ} :
+  differentiable_within_at ℝ arcsin (Ici x) x ↔ x ≠ -1 :=
+begin
+  refine ⟨_, λ h, (has_deriv_within_at_arcsin_Ici h).differentiable_within_at⟩,
+  rintro h rfl,
+  have : sin ∘ arcsin =ᶠ[𝓝[Ici (-1:ℝ)] (-1)] id,
+  { filter_upwards [Icc_mem_nhds_within_Ici ⟨le_rfl, neg_lt_self (@zero_lt_one ℝ _ _)⟩],
+    exact λ x, sin_arcsin' },
+  have := h.has_deriv_within_at.sin.congr_of_eventually_eq this.symm (by simp),
+  simpa using (unique_diff_on_Ici _ _ left_mem_Ici).eq_deriv _ this (has_deriv_within_at_id _ _)
+end
+
+lemma differentiable_within_at_arcsin_Iic {x : ℝ} :
+  differentiable_within_at ℝ arcsin (Iic x) x ↔ x ≠ 1 :=
+begin
+  refine ⟨λ h, _, λ h, (has_deriv_within_at_arcsin_Iic h).differentiable_within_at⟩,
+  rw [← neg_neg x, ← image_neg_Ici] at h,
+  have := (h.comp (-x) differentiable_within_at_id.neg (maps_to_image _ _)).neg,
+  simpa [(∘), differentiable_within_at_arcsin_Ici] using this
+end
+
+lemma differentiable_at_arcsin {x : ℝ} :
+  differentiable_at ℝ arcsin x ↔ x ≠ -1 ∧ x ≠ 1 :=
+⟨λ h, ⟨differentiable_within_at_arcsin_Ici.1 h.differentiable_within_at,
+  differentiable_within_at_arcsin_Iic.1 h.differentiable_within_at⟩,
+  λ h, (has_deriv_at_arcsin h.1 h.2).differentiable_at⟩
+
+@[simp] lemma deriv_arcsin : deriv arcsin = λ x, 1 / sqrt (1 - x ^ 2) :=
```

Comment:
It's kind of amazing that the junk values of `deriv`, `/` and `sqrt` match up to make this statement true.